tests/pylato/test_hamiltonian.py

```diff
@@ -70,31 +120,121 @@ def test_total_energy_happy_path(self):
+        # fake electrostatics (for collinear case)
+        hami.Wi = [1 for i in range(Job.NAtom)]
+
+        # Action
+        result = hami.add_Coulomb_term(Job, 0, 0, U, J, J, dJ, rho)
+
+        # Result
+        assert result == expected_value
 
     def test_add_Coulomb_term_unrecognised_model_error(self):
         Job = InitJob("test_data/JobDef_dcase.json")
         rho = np.eye(10)
 
         hami = Hamiltonian(Job)
+        Job.Def["Hamiltonian"] = "invalid_hami_type"
 
         with pytest.raises(UnimplementedModelError):
-            hami.add_Coulomb_term(0, 0, 1, 1, 1, 1, 1, 1, rho,
-                                  "invalid_hami_type")
+            hami.add_Coulomb_term(Job, 0, 0, 1, 1, 1, 1, rho)
+
+    @pytest.mark.parametrize(
+        ("spin1", "spin2", "expected_value"),
+        [
+            (0, 0, -0.5),
+            (0, 1, complex(-0.5, 0.5)),
+            (1, 0, complex(-0.5, -0.5)),
+            (1, 1, 0.5),
+        ]
+    )
+    def test_spin_onsite_energy_shift(self, spin1, spin2, expected_value):
+        # Setup
+        Job = InitJob("test_data/JobDef_collinear.json")
+
+        # Fake the Stoner I and the spin values for all atoms
+        for atom in range(len(Job.Model.atomic)):
+            atype = Job.AtomType[atom]
+            Job.Model.atomic[atype]['I'] = 1
+            Job.Hamilton.s[0, atom] = 1
+            Job.Hamilton.s[1, atom] = 1
+            Job.Hamilton.s[2, atom] = 1
+
+        # Action
+        result = Job.Hamilton.spin_onsite_energy_shift(Job, spin1, spin2, 0)
+
+        # Result
+        assert result == expected_value
+
+    @pytest.mark.parametrize(
+        ("name", "ii", "jj", "expected_value"),
+        [
+            ("same atom (0) and same spin (0)", 0, 0, 2),
+            ("same atom (1) and same spin (0)", 1, 1, 2),
+            ("same atom (0) and same spin (1)", 2, 2, 2),
+            ("same atom (1) and same spin (1)", 3, 3, 2),
+            ("same atom (0) opposite spin", 0, 2, 1),
+            ("same atom (1) opposite spin", 1, 3, 1),
+            ("different atom and same spin (0)", 0, 1, 0),
+            ("different atom and same spin (1)", 3, 2, 0),
+            ("different atom and opposite spin", 0, 3, 0),
+            ("different atom and opposite spin again", 2, 1, 0),
+        ]
+    )
+    def test_add_H_collinear(self, monkeypatch, name, ii, jj, expected_value):
+        # fake spin_onsite_energy_shift
+        monkeypatch.setattr(
+            'pylato.hamiltonian.Hamiltonian.spin_onsite_energy_shift',
+            fake_spin_onsite_energy_shift
+        )
+
+        # Setup
+        Job = InitJob("test_data/JobDef_collinear.json")
+        # fake electrostatics (Hamilton.Wi)
+        Job.Hamilton.Wi = [1 for i in range(Job.NAtom)]
+
+        # Action
+        result = Job.Hamilton.add_H_collinear(Job, ii, jj)
+
+        # Result
+        assert result == expected_value
+
+    def test_buildH0_one_atom(self):
```

Comment:
@horsfielda I think I just found a mistake in building H0. Before we were adding hopping terms from an atom onto itself, which I believe is incorrect. Right? I've corrected it above (I'll add a comment there too) and added this test to enforce this. Please let me know if I'm incorrect in thinking this though.

Reply:
Ahhh ooops... I've had a closer look at the code and we are explicitly expecting onsite blocks, which should be: `Job.Model.atomic[type]['e']`